src/main/scala/com/github.sstone/amqp/ChannelOwner.scala

```diff
@@ -51,10 +60,9 @@ object ChannelOwner {
           def handleFlow(active: Boolean): Unit = listener ! HandleFlow(active)
         }))
       }
-      case request@Publish(exchange, routingKey, body, properties, mandatory, immediate) => {
+      case request@Publish(_, _, _, _, _, _) => {
```

Comment:
I agree with Mikolaj that isInstanceOf-like things are coding smells and make the code more error-prone so we should avoid them.

On the other hand, I was curious and analyzed a bit the bytecode that we end up having in each case (with IntelliJ-s built-in bytecode viewer):

class TypedPattern {
	case class Point(x: Long, y: Long)
	def makeMatchTyped(a: Any) = {
		a match {
			case p: Point => 5
		}
	}
}

class ContructionPattern {
	case class Point(x: Long, y: Long)
	def makeMatchConstr(a: Any) = {
		a match {
			case p@Point(_, _) => 5
		}
	}
}

The bytecode, only for the methods:
1) Typed pattern
// access flags 0x1
  public makeMatchTyped(Ljava/lang/Object;)I
   L0
    LINENUMBER 9 L0
    ALOAD 1
    ASTORE 2
   L1
    LINENUMBER 10 L1
    ALOAD 2
   ** INSTANCEOF levi/TypedPattern$Point **
    IFEQ L2
    ALOAD 2
    ** CHECKCAST levi/TypedPattern$Point **
    ** INVOKEVIRTUAL levi/TypedPattern$Point.levi$TypedPattern$Point$$$outer ()Llevi/TypedPattern; **
    ALOAD 0
    IF_ACMPNE L2
    ICONST_5
    ISTORE 3
   L3
    LINENUMBER 9 L3
    ILOAD 3
    IRETURN
   L2
   FRAME APPEND [java/lang/Object]
    NEW scala/MatchError
    DUP
    ALOAD 2
    INVOKESPECIAL scala/MatchError.<init> (Ljava/lang/Object;)V
    ATHROW
   L4
    LOCALVARIABLE this Llevi/TypedPattern; L0 L4 0
    LOCALVARIABLE a Ljava/lang/Object; L0 L4 1
    MAXSTACK = 3
    MAXLOCALS = 4

2) Construction pattern

  // access flags 0x1
  public makeMatchConstr(Ljava/lang/Object;)I
   L0
    LINENUMBER 9 L0
    ALOAD 1
    ASTORE 2
   L1
    LINENUMBER 10 L1
    ALOAD 2
   ** INSTANCEOF levi/ContructionPattern$Point **
    IFEQ L2
    ALOAD 2
   ** CHECKCAST levi/ContructionPattern$Point **
   ** INVOKEVIRTUAL levi/ContructionPattern$Point.levi$ContructionPattern$Point$$$outer **()Llevi/ContructionPattern;
    ALOAD 0
    IF_ACMPNE L2
    ICONST_5
    ISTORE 3
   L3
    LINENUMBER 9 L3
    ILOAD 3
    IRETURN
   L2
   FRAME APPEND [java/lang/Object]
    NEW scala/MatchError
    DUP
    ALOAD 2
    INVOKESPECIAL scala/MatchError.<init> (Ljava/lang/Object;)V
    ATHROW
   L4
    LOCALVARIABLE this Llevi/ContructionPattern; L0 L4 0
    LOCALVARIABLE a Ljava/lang/Object; L0 L4 1
    MAXSTACK = 3
    MAXLOCALS = 4

Apart from the classnames, it compiles to the same bytecode (calling instanceof and  checkcast instructions!)

Answering the last question, * request: Publish * looks neater request@Publish(_, _, _, _, _, _) * :)

All in all, if it compiles to the same bytecode, I would probably chose the first, but anyway, we can ask Martin about the drawback of each at Scala Days in Amsterdam.